### ISSUE TITLE
Prometheus Deployment Not Working on PKS

### ISSUE THREAD

**Issue by kacole2** [CNA-Tech/Apps-on-PKS#12]
The Prometheus deployment authored by @pdaigle isn't working on PKS.

Everything works up until the deployment of Prometheus
```
helm install stable/prometheus-operator --name prometheus --namespace monitoring -f custom.yaml
```
When performing this command it has varied results. Sometimes errors appear about CRDs that are currently in process of being deleted. After deleting the CRDs following the [prometheus helm deployment deletion instructions](https://github.com/helm/charts/tree/master/stable/prometheus-operator#uninstalling-the-chart) and retrying, the installation may work. It usually takes a few tries. I have not tried installing the CRDs first and then performing the HELM deployment. 

Helm deployment is not working because all pods are not coming up.
```
root@PKS-client-DEMO:~# kubectl get pods -n monitoring
NAME                                                   READY   STATUS              RESTARTS   AGE
prometheus-grafana-f9dd85d7b-5gvbg                     0/2     Init:0/2            0          64m
prometheus-kube-state-metrics-746bd49df8-ljjnr         0/1     ContainerCreating   0          64m
prometheus-prometheus-node-exporter-2bnkw              1/1     Running             0          64m
prometheus-prometheus-node-exporter-gspg4              1/1     Running             0          64m
prometheus-prometheus-node-exporter-hgk4j              1/1     Running             0          64m
prometheus-prometheus-oper-operator-7d47cdf47f-4l277   0/1     ContainerCreating   0          64m
```

Getting errors that the `pod sandbox changed` for pods that aren't coming up. From my google-fu, this looks to be a CNI or DNS issue ([example1](https://github.com/kubernetes/kubeadm/issues/507#issuecomment-354462050), [example2](https://stackoverflow.com/questions/50744371/pod-creation-stuck-in-containercreating-state/50800163), [example3](https://github.com/kubernetes/kops/issues/4327))
```
root@PKS-client-DEMO:~# kubectl describe pod prometheus-prometheus-oper-operator-7d47cdf47f-4l277 -n monitoring
Name:               prometheus-prometheus-oper-operator-7d47cdf47f-4l277
Namespace:          monitoring
Priority:           0
PriorityClassName:  <none>
Node:               17ee183b-fee0-48c7-8b08-0d05050f2c05/172.23.5.6
Start Time:         Tue, 30 Apr 2019 10:32:07 -0700
Labels:             app=prometheus-operator-operator
                    chart=prometheus-operator-5.5.0
                    heritage=Tiller
                    pod-template-hash=7d47cdf47f
                    release=prometheus
Annotations:        <none>
Status:             Pending
IP:
Controlled By:      ReplicaSet/prometheus-prometheus-oper-operator-7d47cdf47f
Containers:
  prometheus-operator:
    Container ID:
    Image:         quay.io/coreos/prometheus-operator:v0.29.0
    Image ID:
    Port:          8080/TCP
    Host Port:     0/TCP
    Args:
      --kubelet-service=kube-system/prometheus-prometheus-oper-kubelet
      --logtostderr=true
      --crd-apigroup=monitoring.coreos.com
      --localhost=127.0.0.1
      --prometheus-config-reloader=quay.io/coreos/prometheus-config-reloader:v0.29.0
      --config-reloader-image=quay.io/coreos/configmap-reload:v0.0.1
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from prometheus-prometheus-oper-operator-token-p2pxz (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  prometheus-prometheus-oper-operator-token-p2pxz:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  prometheus-prometheus-oper-operator-token-p2pxz
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                  Age                 From                                           Message
  ----     ------                  ----                ----                                           -------
  Warning  FailedCreatePodSandBox  49s (x16 over 60m)  kubelet, 17ee183b-fee0-48c7-8b08-0d05050f2c05  Failed create pod sandbox: rpc error: code = DeadlineExceeded desc = context deadline exceeded
  Normal   SandboxChanged          48s (x16 over 60m)  kubelet, 17ee183b-fee0-48c7-8b08-0d05050f2c05  Pod sandbox changed, it will be killed and re-created.
```

persistent volumes are being created and bound as expected:
```
root@PKS-client-DEMO:~# kubectl get pvc -n monitoring
NAME                 STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
prometheus-grafana   Bound    pvc-e024f1fc-6b6d-11e9-9bfe-005056801138   1Gi        RWO            default        76m
```

The directions also call for using NGNIX as our Ingress controller. Is there a reason we aren't using [Contour](https://github.com/heptio/contour)? Or perhaps a second example showing a NSX-T Load Balancer? 

I'm not 100% sure this is correct, but is the NGINX Ingress Controller automatically creating the LoadBalancer in NSX-T? The only reason I can potentially get there is from the findings below:

I have also tried all these directions on a standard vSphere cluster created by kubeadm without NSX-T and I'm able to finish the deployment. I'm only able to get to the Grafana dashboard by exposing the service as NodePort as well.

When looking at the default services after deploying on my non-PKS cluster, I see that NGINX is creating a type LoadBalancer which won't happen. Assuming this is where NSX-T should come in:
```
[kcoleman@k8s-01-master01 ~]$ kubectl get svc
NAME                                       TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)                      AGE
kubernetes                                 ClusterIP      10.96.0.1       <none>        443/TCP                      61m
quickstart-nginx-ingress-controller        LoadBalancer   10.110.22.216   <pending>     80:30582/TCP,443:32069/TCP   55m
quickstart-nginx-ingress-default-backend   ClusterIP      10.110.87.189   <none>        80/TCP                       55m
```

/assign @pdaigle

**Comment by kacole2** [comment]
Just tried again in the vRA Lab and cannot get the deployments to create their replica sets. 

```
ubuntu@ubuntu:~/prometheus$ kubectl get pods -n monitoring
No resources found.
```

When looking at the deployment:
```
ubuntu@ubuntu:~/prometheus$ kubectl describe deployment prometheus-prometheus-oper-operator -n monitoring
Name:                   prometheus-prometheus-oper-operator
Namespace:              monitoring
CreationTimestamp:      Tue, 30 Apr 2019 13:46:06 -0700
Labels:                 app=prometheus-operator-operator
                        chart=prometheus-operator-5.5.0
                        heritage=Tiller
                        release=prometheus
Annotations:            deployment.kubernetes.io/revision: 1
Selector:               app=prometheus-operator-operator,release=prometheus
Replicas:               1 desired | 0 updated | 0 total | 0 available | 1 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           app=prometheus-operator-operator
                    chart=prometheus-operator-5.5.0
                    heritage=Tiller
                    release=prometheus
  Service Account:  prometheus-prometheus-oper-operator
  Containers:
   prometheus-operator:
    Image:      quay.io/coreos/prometheus-operator:v0.29.0
    Port:       8080/TCP
    Host Port:  0/TCP
    Args:
      --kubelet-service=kube-system/prometheus-prometheus-oper-kubelet
      --logtostderr=true
      --crd-apigroup=monitoring.coreos.com
      --localhost=127.0.0.1
      --prometheus-config-reloader=quay.io/coreos/prometheus-config-reloader:v0.29.0
      --config-reloader-image=quay.io/coreos/configmap-reload:v0.0.1
    Environment:  <none>
    Mounts:       <none>
  Volumes:        <none>
Conditions:
  Type             Status  Reason
  ----             ------  ------
  Progressing      True    NewReplicaSetCreated
  Available        False   MinimumReplicasUnavailable
  ReplicaFailure   True    FailedCreate
OldReplicaSets:    <none>
NewReplicaSet:     prometheus-prometheus-oper-operator-7d47cdf47f (0/1 replicas created)
Events:
  Type    Reason             Age    From                   Message
  ----    ------             ----   ----                   -------
  Normal  ScalingReplicaSet  3m24s  deployment-controller  Scaled up replica set prometheus-prometheus-oper-operator-7d47cdf47f to 1
```

Same thing happens with the Deployment of NGINX:
```
ubuntu@ubuntu:~/prometheus$ kubectl describe deployment quickstart-nginx-ingress-controller
Name:                   quickstart-nginx-ingress-controller
Namespace:              default
CreationTimestamp:      Tue, 30 Apr 2019 13:45:03 -0700
Labels:                 app=nginx-ingress
                        chart=nginx-ingress-1.6.0
                        component=controller
                        heritage=Tiller
                        release=quickstart
Annotations:            deployment.kubernetes.io/revision: 1
Selector:               app=nginx-ingress,component=controller,release=quickstart
Replicas:               1 desired | 0 updated | 0 total | 0 available | 1 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  1 max unavailable, 1 max surge
Pod Template:
  Labels:           app=nginx-ingress
                    component=controller
                    release=quickstart
  Service Account:  quickstart-nginx-ingress
  Containers:
   nginx-ingress-controller:
    Image:       quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.24.1
    Ports:       80/TCP, 443/TCP
    Host Ports:  0/TCP, 0/TCP
    Args:
      /nginx-ingress-controller
      --default-backend-service=default/quickstart-nginx-ingress-default-backend
      --election-id=ingress-controller-leader
      --ingress-class=nginx
      --configmap=default/quickstart-nginx-ingress-controller
    Liveness:   http-get http://:10254/healthz delay=10s timeout=1s period=10s #success=1 #failure=3
    Readiness:  http-get http://:10254/healthz delay=10s timeout=1s period=10s #success=1 #failure=3
    Environment:
      POD_NAME:        (v1:metadata.name)
      POD_NAMESPACE:   (v1:metadata.namespace)
    Mounts:           <none>
  Volumes:            <none>
Conditions:
  Type             Status  Reason
  ----             ------  ------
  Available        True    MinimumReplicasAvailable
  ReplicaFailure   True    FailedCreate
OldReplicaSets:    <none>
NewReplicaSet:     quickstart-nginx-ingress-controller-6445784d55 (0/1 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  11m   deployment-controller  Scaled up replica set quickstart-nginx-ingress-controller-6445784d55 to 1
```

**Comment by kacole2** [comment]
After Francis updated the PKS lab to 1.4 I was able to install prometheus following the directions, but had issues on getting the initial installation. 

When installing prometheus, the helm chart exports this issue.
```
root@PKS-client-DEMO:~/kenny-prom# helm install stable/prometheus-operator --name prometheus --namespace monitoring -f custom.yaml
Error: validation failed: [unable to recognize "": no matches for kind "Alertmanager" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "Prometheus" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "PrometheusRule" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1", unable to recognize "": no matches for kind "ServiceMonitor" in version "monitoring.coreos.com/v1"]
```

I've seen this before when using Istio. It's a CRD issue but it's not always apparent. I had to delete the CRDs and re-run the installation and it installed just fine. A simple work around will be installing the CRDs first and then running the helm installation but it will need to be tested.

**Comment by kacole2** [comment]
This has been solved with #13
